packages/client-utils/src/vstorage-client.js

```diff
@@ -0,0 +1,489 @@
+import {
+  QueryChildrenRequest,
+  QueryChildrenResponse,
+  QueryDataRequest,
+  QueryDataResponse,
+} from '@agoric/cosmic-proto/agoric/vstorage/query.js';
+import { encodeHex } from '@agoric/internal/src/hex.js';
+import { decodeBase64 } from '@endo/base64';
+import { Fail, q as quote } from '@endo/errors';
+
+/**
+ * @import {MinimalNetworkConfig} from '@agoric/client-utils/src/network-config';
+ * @import {JsonSafe} from '@agoric/cosmic-proto';
+ * @import {AbciQueryResponse} from '@cosmjs/tendermint-rpc';
+ *
+ * @typedef {bigint} Height;
+ *
+ * @typedef {JsonRpcResponse<{response: AbciQueryResponse}>} AbciResponse;
+ *
+ * @typedef {JsonRpcResponse<{
+ *  sync_info: {
+ *    catching_up: boolean;
+ *    latest_block_height: string;
+ *  }
+ * }>} NodeStatusResponse;
+ *
+ *
+ * @typedef {{
+ *  blockHeight: string;
+ *  values: Array<string>;
+ * }} StreamCell
+ */
+
+/**
+ * @template {any} T
+ * @typedef {{
+ *  id: number;
+ *  jsonrpc: string;
+ *  result: JsonSafe<T>
+ * }} JsonRpcResponse;
+ */
+
+/**
+ * @template {any} T
+ * @typedef {{
+ *  latest(height?: Height): Promise<T>;
+ * }} Topic<T>
+ */
+
+/**
+ * @template {any} T
+ * @typedef {Topic<T> & {
+ *  iterate(minimum?: Height, maximum?: Height): AsyncIterable<Update<T>>;
+ *  reverseIterate(maximum?: Height, minimum?: Height): AsyncIterable<Update<T>>;
+ * }} StreamTopic<T>
+ */
+
+/**
+ * @template {any} T
+ * @typedef {{
+ *  blockHeight: Height;
+ *  value: T;
+ * }} Update<T>
+ */
+
+export const INVALID_HEIGHT_ERROR_MESSAGE = 'Invalid height';
+export const INVALID_RPC_ADDRESS_ERROR_MESSAGE =
+  'No valid RPC address found in config';
+
+const MINIMUM_HEIGHT = 1n;
+
+export const PATH_PREFIX = '/agoric.vstorage.Query';
+
+export const PATHS = /** @type {const} */ ({
+  CHILDREN: 'Children',
+  DATA: 'Data',
+});
+
+const codecs = /** @type {const} */ ({
+  [PATHS.CHILDREN]: {
+    request: QueryChildrenRequest,
+    response: QueryChildrenResponse,
+  },
+  [PATHS.DATA]: {
+    request: QueryDataRequest,
+    response: QueryDataResponse,
+  },
+});
+
+/**
+ * @param {string} errorMessage
+ * @param {Height} height
+ * @param {PATHS[keyof PATHS]} kind
+ * @param {string} path
+ * @param {MinimalNetworkConfig['rpcAddrs'][0]} [rpcAddress]
+ */
+const createError = (errorMessage, height, kind, path, rpcAddress) =>
```

Comment:
Consider `makeVstorageError`. We have a prevailing convention for `make`, and also have `assert.makeError`, and this forces us to clarify that this is specific to the vstorage domain.